pkg/reconciler/v1alpha1/taskrun/entrypoint/entrypoint.go

```diff
@@ -76,116 +83,151 @@ func (c *Cache) set(sha string, ep []string) {
+	ref, err := name.ParseReference(image, name.WeakValidation)
 	if err != nil {
-		return "", fmt.Errorf("couldn't get digest hash for image %s: %v", image, err)
+		return nil, fmt.Errorf("Failed to parse image %s: %v", image, err)
 	}
-	// Parse Digest Hash struct into sha string
-	digest := fmt.Sprintf("%s%s%s", image, digestSeparator, digestHash.String())
-	cache.set(image, []string{digest})
-
-	return digest, nil
-}
-
-// RedirectSteps will modify each of the steps/containers such that
-// the binary being run is no longer the one specified by the Command
-// and the Args, but is instead the entrypoint binary, which will
-// itself invoke the Command and Args, but also capture logs.
-func RedirectSteps(steps []corev1.Container) error {
-	for i := range steps {
-		step := &steps[i]
-		e, err := getEnvVar(step.Command, step.Args)
+	var kc authn.Keychain
+	serviceAccountName := build.Spec.ServiceAccountName
+	if serviceAccountName != "" && serviceAccountName != "default" {
+		sa, err := kubeclient.CoreV1().ServiceAccounts(build.Namespace).Get(serviceAccountName, metav1.GetOptions{})
 		if err != nil {
-			return fmt.Errorf("couldn't get env var for entrypoint: %s", err)
+			return nil, fmt.Errorf("Failed to get service account %s: %v", serviceAccountName, err)
 		}
-		step.Command = []string{BinaryLocation}
-		step.Args = []string{}
-
-		step.Env = append(step.Env, corev1.EnvVar{
-			Name:  JSONConfigEnvVar,
-			Value: e,
+		if len(sa.ImagePullSecrets) == 0 {
+			return nil, fmt.Errorf("No ImagePullSecret for service account %s: %v", serviceAccountName, err)
+		}
+		for _, secret := range sa.ImagePullSecrets {
+			_, err = kubeclient.CoreV1().Secrets(build.Namespace).Get(secret.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, fmt.Errorf("Failed to get ImagePullSecret for service account %s: %v", serviceAccountName, err)
+			}
+
+			kc, err = k8schain.New(kubeclient, k8schain.Options{
+				ImagePullSecrets: []string{secret.Name},
+				Namespace:        build.Namespace,
+			})
+			if err != nil {
+				return nil, fmt.Errorf("Failed to create k8schain: %v", err)
+			}
+		}
+	} else {
+		kc, err = k8schain.New(kubeclient, k8schain.Options{
+			Namespace:          build.Namespace,
+			ServiceAccountName: "default",
 		})
-		step.VolumeMounts = append(step.VolumeMounts, toolsMount)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to create k8schain: %v", err)
+		}
 	}
-	return nil
-}
 
-func getRemoteImage(image string) (v1.Image, error) {
-	// verify the image name, then download the remote config file
-	ref, err := name.ParseReference(image, name.WeakValidation)
-	if err != nil {
-		return nil, fmt.Errorf("couldn't parse image %s: %v", image, err)
-	}
-	img, err := remote.Image(ref, remote.WithAuthFromKeychain(authn.DefaultKeychain))
+	// this will first try to authenticate using the k8schain,
+	// then fall back to the google keychain,
+	// then fall back to anonymous
```

Comment:
i think the docs are a bit light on how this works, could we add a bit more? (i think right now they just say you should provide a serviceaccount that has the right credentials, but this logic looks like it'll do a bit more)

Reply:
will add more